cs-property-service/src/main/java/org/egov/cpt/service/DemandGenerationService.java

```diff
@@ -0,0 +1,131 @@
+			propertyCriteria.setPropertyId(property.getId());
+			List<RentDemand> rentDemandList = propertyRepository.getPropertyRentDemandDetails(propertyCriteria);
+			// System.out.println(rentDemandList);
+
+			if (!rentDemandList.isEmpty()) {
+
+				Comparator<RentDemand> compare = Comparator.comparing(RentDemand::getGenerationDate);
+				Optional<RentDemand> collectionDemand = rentDemandList.stream().min(compare);
+
+				int currentYear = LocalDate.now().getYear();
+				List<String> dateList = rentDemandList.stream().map(r -> r.getGenerationDate())
+						.map(date -> new Date(date).toInstant().atZone(ZoneId.systemDefault()).toLocalDate().toString())
+						// .filter(d -> d.getYear()< currentYear)
+						// .map(local -> local.getMonth().toString())
+						.collect(Collectors.toList());
+
+				// System.out.println(monthList);
+				if (demandCriteria.getDate().isEmpty()) {
+					String currentDate = LocalDate.now().toString();
+
+					if (!dateList.contains(currentDate)) {
+						// generate demand
+						generateRentDemand(property, collectionDemand.get(), currentYear);
+					}
+				} else {
+					String date = LocalDate.parse(demandCriteria.getDate(), FORMATTER).toString();
+					if (!dateList.contains(date)) {
+						// generate demand
+						generateRentDemand(property, collectionDemand.get(), currentYear);
+					}
+				}
+			} else {
+				// generate demand
+				// generateRentDemand(property, collectionDemand);
+			}
+		});
+
+	}
+
+	private void generateRentDemand(Property property, RentDemand collectionDemand, int currentYear) {
+
+		int oldYear = new Date(collectionDemand.getGenerationDate()).toInstant().atZone(ZoneId.systemDefault())
+				.toLocalDate().getYear();
+
+		RentDemand rentDemand = new RentDemand();
+		String rendDemandId = UUID.randomUUID().toString();
+				rentDemand.setId(rendDemandId);
+		rentDemand.setPropertyId(property.getId());
+
+		rentDemand.setGenerationDate(LocalDateTime.now().atZone(ZoneId.systemDefault()).toInstant().toEpochMilli());
+
+		Double collectionPrincipal = collectionDemand.getCollectionPrincipal();
+		while (oldYear < currentYear) {
+			collectionPrincipal = (collectionPrincipal
+					* (100 + property.getPropertyDetails().getRentIncrementPercentage())) / 100;
+			oldYear = oldYear + property.getPropertyDetails().getRentIncrementPeriod();
+		}
+
+		rentDemand.setCollectionPrincipal(collectionPrincipal);
+		rentDemand.setRemainingPrincipal(rentDemand.getCollectionPrincipal());
+		rentDemand.setInterestSince(rentDemand.getGenerationDate());
+		//rentDemand.setTenantId(property.getTenantId());
+		// rentDemand.setAuditDetails();
+		
+		//property.setDemands(Collections.singletonList(rentDemand));
+		
+		log.info("rend demand id: "+ rendDemandId);
+		log.info("collection principal: "+ collectionPrincipal);
+		
+		if(property.getPayments() != null && property.getRentAccount() != null) {
+			rentCollectionService.settle(property.getDemands(), property.getPayments(), property.getRentAccount(),
```

Comment:
`settle` method gives you `RentCollection` objects. It will also modify the `rentAccount`, `demands` and `payments` objects sent to it. We need to set them to the property and push the property to be saved on to kafka.